torch/_inductor/codegen/common.py

```diff
@@ -870,3 +871,57 @@ class OptimizationContext:
     is_load_uint8_as_float: bool = False
     # Store float32 value as uint8
     is_store_float_as_uint8: bool = False
+
+
+def can_use_32bit_indexing(
+    node_schedule, numel: sympy.Expr, reduction_numel: sympy.Expr
+) -> bool:
+    from ..ir import Buffer, ConstantBuffer, FixedLayout
+    from ..scheduler import BaseSchedulerNode
+
+    buffer_names = set()
+    for node in node_schedule:
+        if not isinstance(node, BaseSchedulerNode):
+            continue
+
+        buffer_names.update(node.get_names())
+        buffer_names.update(node.used_buffer_names())
+
+    # Get buffers objects
+    def _get_buffer(name: str) -> Buffer:
+        if name in V.graph.name_to_buffer:
+            return V.graph.name_to_buffer[name]
+        elif name in V.graph.graph_inputs:
+            return V.graph.graph_inputs[name]
+        elif name in V.graph.constants:
+            data = V.graph.constants[name]
+            return ConstantBuffer(
+                name,
+                FixedLayout(
+                    data.device, data.dtype, *V.graph.static_sizes_strides(data)
+                ),
+            )
+        raise RuntimeError(f"Failed to find buffer matching name {name}")
+
+    buffers = [_get_buffer(name) for name in buffer_names]
+
+    # In theory we can separately check xnumel and rnumel are <= int_max
+    # but some indexers do use the full linear index so we need to be
+    # conservative here.
+    total_numel = numel * reduction_numel
```

Comment:
Note that you need to take the strides into account. The end of the tensor is at `max(strides[i] * size[i] for i in range(n)) - offset`

Reply:
`buf_sizes` below does count the strides.